shell/platform/android/io/flutter/view/AccessibilityViewEmbedder.java

```diff
@@ -220,6 +225,122 @@ private void copyAccessibilityFields(AccessibilityNodeInfo input, AccessibilityN
         }
     }
 
+    /**
+     * Delegates an AccessibilityNodeProvider#requestSendAccessibilityEvent from the AccessibilityBridge to the embedded
+     * view.
+     *
+     * @return True if the event was sent.
+     */
+    public boolean requestSendAccessibilityEvent(View embeddedView, View eventOrigin, AccessibilityEvent event) {
+        AccessibilityEvent translatedEvent = AccessibilityEvent.obtain(event);
+        Long originPackedId = reflectionAccessors.getRecordSourceNodeId(event);
+        if (originPackedId == null) {
+            return false;
+        }
+        int originVirtualId = ReflectionAccessors.getVirtualNodeId(originPackedId);
+        Integer flutterId = originToFlutterId.get(new ViewAndId(embeddedView, originVirtualId));
+        if (flutterId == null) {
+            return false;
+        }
+        translatedEvent.setSource(rootAccessibilityView, flutterId);
+        translatedEvent.setClassName(event.getClassName());
+        translatedEvent.setPackageName(event.getPackageName());
+
+        for (int i = 0; i < translatedEvent.getRecordCount(); i++) {
+            AccessibilityRecord record = translatedEvent.getRecord(i);
+            Long recordOriginPackedId = reflectionAccessors.getRecordSourceNodeId(record);
+            if (recordOriginPackedId == null) {
+                return false;
+            }
+            int recordOriginVirtualID = ReflectionAccessors.getVirtualNodeId(recordOriginPackedId);
+            Integer recordFlutterId = originToFlutterId.get(new ViewAndId(embeddedView, recordOriginVirtualID));
+            if (recordFlutterId == null) {
```

Comment:
(subjective, not in scope for this change, doesn't need to be resolved): I think relying on `null` as a silent error condition like here is dangerous. It's hard to know when to look for it. Especially on a type like `Integer` that tends to get equated with the non nullable `int`. I'd rather there was a constant for `INVALID` that callers were expected to check for, and the method returned an `int` instead.

Reply:
FYI, I actually asked @amirh to switch from a magic number to null.  Can you elaborate on why null would be bad?  Null is often used to express non-existence, whereas a value definitely exists but has some special meaning that is different than all other numbers.  To me, the latter seemed more likely to confuse people, especially if you ever do a debug dump and see just the magic number instead of `null`.  Are there likely situations where `null` would create issues?

Reply:
Changed to check `containsKey`'s return value.

Reply:
Sure thing. I do think this is subjective so the patch as it is is also fine. And I agree magic numbers are unclear, so that's why I was suggesting a named constant for the error case specifically.

In general I think all deliberate uses of `null` are somewhat dangerous and prefer that non-nullable be default and enforced with static analysis if possible, so that's mostly what's prompting my suggestion here. ["Null References: The Billion Dollar Mistake"](https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare) is a really good talk going into deeper issues with it. Items 54-55 in *Effective Java* also touch on problems with `null`.

My main dislike of deliberate null returns comes down to a combination of two things.

- Nullable values are invisible. It's up to the programmer calling the method to proactively figure out that the return can be null and handle it appropriately, and usually the only tools at their disposal are untrustworthy manually added markers like annotations, comments, etc which can all be out of date, missing, or hard to notice in the first place. The only way to really know is to read the implementation details of every method that returns a potentially nullable type.
- Missing `null` means getting an NPE and a program crash, so checking for it is critically important at the same time as being really hard to know when to guard against. NPEs can crop up far removed from the original nullable getter and be generally hard to diagnose.

It's basically breaking down the type system. If any type can be either itself or `null`, then all code using nullable values needs to first basically do type checking to make sure the value is actually the thing that they're expecting before relying on it like it's an instance of its type. With this `Integer` case specifically I'd be especially worried about somebody not realizing that this method is nullable and just trying to cast it to `int` or even comparing it with another `int` with `==`, which would cause an NPE.

These concerns outweigh the semantic harmony of literally returning "nothing" to match that we found "nothing" for me. I'd rather we return a status code labeled "NOT_FOUND" or similar for clients to handle.

`Optional<int>` is probably the better way to go in a lot of these kinds of cases. It points out that it's possible to get "nothing" back from a method in a deliberate way that's impossible to miss while coding. But I don't think it's available in Android APIs before 24.